sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -182,6 +183,10 @@ case class DataSource(
       throw new AnalysisException(
         s"Unable to infer schema for $format. It must be specified manually.")
     }
+
+    SchemaUtils.checkSchemaColumnNameDuplication(
+      dataSchema, "datasource", sparkSession.sessionState.conf.caseSensitiveAnalysis)
```

Comment:
IIRC we can not persistent data with duplicated column names in schema, so should be fine